File: src/js-single-thread/processor.js
const {performance} = require("perf_hooks");

class ImageProcessor {
    /**
    * @param {Buffer} - raw pixel data
    * @param {number} width - Image width
    * @param {number} height - Image height
    * @returns {Buffer} Grayscale image data
    */
    toGrayScale(imageData , width , height) {
        const pixels = width * height;
        const output = Buffer.alloc(imageData.length);

        for (let i = 0 ; i < pixels ; i++) {
            const offset = i * 4;
            const r = imageData[offset];
            const g = imageData[offset + 1];
            const b = imageData[offset + 2];
            const a = imageData[offset + 3];

            //Luminosity method -most accurate for human perception-
            const gray = Math.floor(0.299 * r + 0.587 * g + 0.114 * b);

            output[offset] = gray;
            output[offset + 1] = gray;
            output[offset + 2] = gray;
            output[offset + 3] = a; 
        }

        return output;
    }
    /*
    * @param {Buffer} imageData - Raw pixel data
   * @param {number} width - Image width
   * @param {number} height - Image height
   * @param {number} radius - Blur radius (default: 5)
   * @returns {Buffer} Blurred image data
   * */
    applyBlur(imageData , width , height , raduis = 5) {
        const output = Buffer.alloc(imageData.length);
        const diameter = raduis * 2 + 1;
        const kernelSize = diameter * diameter;

        for (let y = 0 ; y < height ; y++) {
            for (let x = 0 ; x < width ; x++) {
                let r = 0 , g = 0 , b = 0 , a = 0;
                let count = 0;

                for (let ky = -raduis ; ky <= raduis ; ky++) {
                    for (let kx = -raduis ; kx <= raduis ; kx++) {
                        const px = x + kx;
                        const py = y + ky;

                        if (px >= 0 && px < width && py >= 0 && py < height) {
                            const offset = (py * width + px) * 4;
                            r += imageData[offset];
                            g += imageData[offset + 1];
                            b += imageData[offset + 2];
                            a += imageData[offset + 3];
                            count++;
                        }
                    }
                }

                const offset = (y * width + x) * 4 ;
                output[offset] = Math.floor(r / count);
                output[offset + 1] = Math.floor(g / count);
                output[offset + 2] = Math.floor(b / count);
                output[offset + 3] = Math.floor(a / count); 
            }
        }
        return output;
    }

    process(imageData , width , height , blurRaduis = 5) {
        console.log(`[Single Thread] processing ${width}x${height} image...`);
        const startTime = performance.now();
        
        const grayScale = this.toGrayScale(imageData , width , height);
        const blurred = this.applyBlur(grayScale , height , width , blurRaduis);
        
        const duration = (performance.now() - startTime) / 1000;
        console.log(`[Single Thread] Completed in ${duration}s`);
        return {data: blurred , duration};
    }
}

module.exports = ImageProcessor;

File: src/cpp-addon/processor.cpp
// src/cpp-addon/processor.cpp
/**
 * C++ Image Processing Implementation
 * Same algorithms as JavaScript version, but in C++ for speed
 */

#include "processor.h"
#include <cmath>
#include <algorithm>
#include <cstring>
#include <stddef.h>

/**
 * Convert RGB image to grayscale using luminosity method
 * Formula: 0.299*R + 0.587*G + 0.114*B
 * 
 * @param input  - Input image buffer (RGBA format)
 * @param output - Output image buffer (RGBA format)
 * @param width  - Image width in pixels
 * @param height - Image height in pixels
 */
void ImageProcessor::toGrayscale(
    const uint8_t* input,
    uint8_t* output,
    int width,
    int height
) {
    // Total number of pixels
    int pixels = width * height;
    
    // Process each pixel
    // Note: We use a pointer approach for speed
    for (int i = 0; i < pixels; i++) {
        // Calculate offset in buffer (each pixel = 4 bytes: RGBA)
        int offset = i * 4;
        
        // Extract RGB values
        uint8_t r = input[offset];
        uint8_t g = input[offset + 1];
        uint8_t b = input[offset + 2];
        uint8_t a = input[offset + 3];
        
        // Calculate grayscale using luminosity method
        // This is optimized: using integer arithmetic instead of float
        // 0.299 ≈ 77/256, 0.587 ≈ 150/256, 0.114 ≈ 29/256
        uint8_t gray = static_cast<uint8_t>(
            (77 * r + 150 * g + 29 * b) >> 8  // Bit shift instead of division!
        );
        
        // Set output pixel (grayscale for RGB, keep alpha)
        output[offset]     = gray;
        output[offset + 1] = gray;
        output[offset + 2] = gray;
        output[offset + 3] = a;
    }
}


/**
 * Apply box blur filter
 * 
 * @param input  - Input image buffer (RGBA format)
 * @param output - Output image buffer (RGBA format)
 * @param width  - Image width in pixels
 * @param height - Image height in pixels
 * @param radius - Blur radius (higher = more blur)
 */
void ImageProcessor::applyBlur(
    const uint8_t* input,
    uint8_t* output,
    int width,
    int height,
    int radius
) {
    // For each pixel in the image
    for (int y = 0; y < height; y++) {
        for (int x = 0; x < width; x++) {
            
            // Accumulators for each channel
            int r = 0, g = 0, b = 0, a = 0;
            int count = 0;
            
            // Iterate over blur kernel (box around current pixel)
            for (int ky = -radius; ky <= radius; ky++) {
                for (int kx = -radius; kx <= radius; kx++) {
                    
                    // Calculate neighbor pixel coordinates
                    int px = x + kx;
                    int py = y + ky;
                    
                    // Check if neighbor is within image bounds
                    if (px >= 0 && px < width && py >= 0 && py < height) {
                        // Calculate offset in buffer
                        int offset = (py * width + px) * 4;
                        
                        // Accumulate channel values
                        r += input[offset];
                        g += input[offset + 1];
                        b += input[offset + 2];
                        a += input[offset + 3];
                        count++;
                    }
                }
            }
            
            // Calculate output pixel offset
            int offset = (y * width + x) * 4;
            
            // Average the accumulated values
            output[offset]     = static_cast<uint8_t>(r / count);
            output[offset + 1] = static_cast<uint8_t>(g / count);
            output[offset + 2] = static_cast<uint8_t>(b / count);
            output[offset + 3] = static_cast<uint8_t>(a / count);
        }
    }
}

void ImageProcessor::process(const uint8_t* input,uint8_t* output,int width,int height,int blurRadius) {
    size_t bufferSize = width * height * 4;
    uint8_t* temp = new uint8_t[bufferSize];
    toGrayscale(input, temp, width, height);
    applyBlur(temp, output, width, height, blurRadius);
    delete[] temp;
}
